File: DocumentsSDK/DocumentPdf/src/main/java/com/proxy/service/document/pdf/core/PdfiumCore.java
//package com.proxy.service.document.pdf.core;
//
//import android.content.Context;
//import android.graphics.Bitmap;
//import android.graphics.Point;
//import android.graphics.RectF;
//import android.os.ParcelFileDescriptor;
//import android.util.Log;
//import android.view.Surface;
//
//import com.proxy.service.document.base.pdf.info.MetaData;
//import com.proxy.service.document.base.pdf.info.PageSize;
//
//import java.io.FileDescriptor;
//import java.io.IOException;
//import java.lang.reflect.Field;
//import java.util.ArrayList;
//import java.util.List;
//
//public class PdfiumCore {
//    private static final String TAG = PdfiumCore.class.getName();
//    private static final Class FD_CLASS = FileDescriptor.class;
//    private static final String FD_FIELD_NAME = "descriptor";
//
//    static {
//        try {
//            System.loadLibrary("c++_shared");
//            System.loadLibrary("modpng");
//            System.loadLibrary("modft2");
//            System.loadLibrary("modpdfium");
//            System.loadLibrary("jniPdfium");
//        } catch (UnsatisfiedLinkError e) {
//            Log.e(TAG, "Native libraries failed to load - " + e);
//        }
//    }
//
//    private native long nativeOpenDocument(int fd, String password);
//
//    private native long nativeOpenMemDocument(byte[] data, String password);
//
//    /**
//     * 关闭对应文档
//     *
//     * @param doc_hand 文档指针
//     */
//    public native void nativeCloseDocument(long doc_hand);
//
//    /**
//     * 获取对应文档内的页面数量
//     *
//     * @param doc_hand 文档指针
//     */
//    public native int nativeGetPageCount(long doc_hand);
//
//    /**
//     * 获取对应文档内对应页面的指针
//     *
//     * @param doc_hand  文档指针
//     * @param pageIndex 页面位置
//     */
//    public native long nativeLoadPage(long doc_hand, int pageIndex);
//
//    private native long[] nativeLoadPages(long docPtr, int fromIndex, int toIndex);
//
//    private native void nativeClosePage(long pagePtr);
//
//    private native void nativeClosePages(long[] pagesPtr);
//
//    private native int nativeGetPageWidthPixel(long pagePtr, int dpi);
//
//    private native int nativeGetPageHeightPixel(long pagePtr, int dpi);
//
//    private native int nativeGetPageWidthPoint(long pagePtr);
//
//    private native int nativeGetPageHeightPoint(long pagePtr);
//
//    //private native long nativeGetNativeWindow(Surface surface);
//    //private native void nativeRenderPage(long pagePtr, long nativeWindowPtr);
//    private native void nativeRenderPage(
//            long pagePtr,
//            Surface surface,
//            int dpi,
//            int startX,
//            int startY,
//            int drawSizeHor,
//            int drawSizeVer,
//            boolean renderAnnot
//    );
//
//    private native void nativeRenderPageBitmap(
//            long pagePtr,
//            Bitmap bitmap,
//            int dpi,
//            int startX,
//            int startY,
//            int drawSizeHor,
//            int drawSizeVer,
//            boolean renderAnnot
//    );
//
//    /**
//     * 获取对应文档说明信息
//     *
//     * @param doc_hand 文档指针
//     * @param tag      标签
//     */
//    public native String nativeGetDocumentMetaText(long doc_hand, String tag);
//
//    /**
//     * 获取对应文档目录的第一个子目录指针
//     *
//     * @param doc_hand       文档指针
//     * @param catalogue_hand 文档目录指针
//     */
//    public native Long nativeGetFirstChildBookmark(long doc_hand, Long catalogue_hand);
//
//    /**
//     * 获取对应文档目录的下一个目录指针
//     *
//     * @param doc_hand       文档指针
//     * @param catalogue_hand 文档目录指针
//     */
//    public native Long nativeGetSiblingBookmark(long doc_hand, long catalogue_hand);
//
//    /**
//     * 获取对应文档目录标题
//     *
//     * @param catalogue_hand 文档目录指针
//     */
//    public native String nativeGetBookmarkTitle(long catalogue_hand);
//
//    /**
//     * 获取文档目录对应的页码
//     *
//     * @param doc_hand       文档指针
//     * @param catalogue_hand 文档目录指针
//     */
//    public native long nativeGetBookmarkDestIndex(long doc_hand, long catalogue_hand);
//
//    private native PageSize nativeGetPageSizeByIndex(long docPtr, int pageIndex, int dpi);
//
//    private native long[] nativeGetPageLinks(long pagePtr);
//
//    private native Integer nativeGetDestPageIndex(long docPtr, long linkPtr);
//
//    private native String nativeGetLinkURI(long docPtr, long linkPtr);
//
//    private native RectF nativeGetLinkRectF(long linkPtr);
//
//    /**
//     * 页面坐标映射为屏幕坐标
//     *
//     * @param pagePtr 页面指针
//     * @param startX  显示区域在设备坐标中的左侧像素位置
//     * @param startY  显示区域在设备坐标中的顶部像素位置
//     * @param sizeX   显示区域的宽度(单位：像素)
//     * @param sizeY   显示区域的高度(单位：像素)
//     * @param rotate  页面方向：0（正常）, 1（顺时针旋转90度）, 2（顺时针旋转180度）, 3（顺时针旋转270度）
//     * @param pageX   页面坐标中的X值
//     * @param pageY   页面坐标中的Y值
//     * @return 返回映射坐标
//     */
//    private native Point nativePageCoordsToDevice(
//            long pagePtr,
//            int startX,
//            int startY,
//            int sizeX,
//            int sizeY,
//            int rotate,
//            double pageX,
//            double pageY
//    );
//
//    /* synchronize native methods */
//    private static final Object lock = new Object();
//    private static Field mFdField = null;
//    private int mCurrentDpi;
//
//    public static int getNumFd(ParcelFileDescriptor fdObj) {
//        try {
//            if (mFdField == null) {
//                mFdField = FD_CLASS.getDeclaredField(FD_FIELD_NAME);
//                mFdField.setAccessible(true);
//            }
//
//            return mFdField.getInt(fdObj.getFileDescriptor());
//        } catch (NoSuchFieldException e) {
//            e.printStackTrace();
//            return -1;
//        } catch (IllegalAccessException e) {
//            e.printStackTrace();
//            return -1;
//        }
//    }
//
//
//    /**
//     * Context needed to get screen density
//     */
////    public PdfiumCore(Context ctx) {
////        mCurrentDpi = ctx.getResources().getDisplayMetrics().densityDpi;
//////        Log.d(TAG, "Starting PdfiumAndroid " + BuildConfig.VERSION_NAME);
////    }
//
//    /**
//     * Create new document from file
//     */
//    public PdfDocument newDocument(ParcelFileDescriptor fd) throws IOException {
//        return newDocument(fd, null);
//    }
//
//    /**
//     * Create new document from file with password
//     */
//    public PdfDocument newDocument(ParcelFileDescriptor fd, String password) throws IOException {
//        PdfDocument document = new PdfDocument();
//        document.parcelFileDescriptor = fd;
//        synchronized (lock) {
//            document.mNativeDocPtr = nativeOpenDocument(getNumFd(fd), password);
//        }
//
//        return document;
//    }
//
//    /**
//     * Create new document from bytearray
//     */
//    public PdfDocument newDocument(byte[] data) throws IOException {
//        return newDocument(data, null);
//    }
//
//    /**
//     * Create new document from bytearray with password
//     */
//    public PdfDocument newDocument(byte[] data, String password) throws IOException {
//        PdfDocument document = new PdfDocument();
//        synchronized (lock) {
//            document.mNativeDocPtr = nativeOpenMemDocument(data, password);
//        }
//        return document;
//    }
//
//    /**
//     * Open page and store native pointer in {@link PdfDocument}
//     */
//    public long openPage(PdfDocument doc, int pageIndex) {
//        long pagePtr;
//        synchronized (lock) {
//            pagePtr = nativeLoadPage(doc.mNativeDocPtr, pageIndex);
//            doc.mNativePagesPtr.put(pageIndex, pagePtr);
//            return pagePtr;
//        }
//
//    }
//
//    /**
//     * Open range of pages and store native pointers in {@link PdfDocument}
//     */
//    public long[] openPage(PdfDocument doc, int fromIndex, int toIndex) {
//        long[] pagesPtr;
//        synchronized (lock) {
//            pagesPtr = nativeLoadPages(doc.mNativeDocPtr, fromIndex, toIndex);
//            int pageIndex = fromIndex;
//            for (long page : pagesPtr) {
//                if (pageIndex > toIndex) break;
//                doc.mNativePagesPtr.put(pageIndex, page);
//                pageIndex++;
//            }
//
//            return pagesPtr;
//        }
//    }
//
//    /**
//     * Render page fragment on {@link Surface}.<br>
//     * Page must be opened before rendering.
//     */
//    public void renderPage(PdfDocument doc, Surface surface, int pageIndex,
//                           int startX, int startY, int drawSizeX, int drawSizeY) {
//        renderPage(doc, surface, pageIndex, startX, startY, drawSizeX, drawSizeY, false);
//    }
//
//    /**
//     * Render page fragment on {@link Surface}. This method allows to render annotations.<br>
//     * Page must be opened before rendering.
//     */
//    public void renderPage(PdfDocument doc, Surface surface, int pageIndex,
//                           int startX, int startY, int drawSizeX, int drawSizeY,
//                           boolean renderAnnot) {
//        synchronized (lock) {
//            try {
//                //nativeRenderPage(doc.mNativePagesPtr.get(pageIndex), surface, mCurrentDpi);
//                nativeRenderPage(doc.mNativePagesPtr.get(pageIndex), surface, mCurrentDpi,
//                        startX, startY, drawSizeX, drawSizeY, renderAnnot);
//            } catch (NullPointerException e) {
//                Log.e(TAG, "mContext may be null");
//                e.printStackTrace();
//            } catch (Exception e) {
//                Log.e(TAG, "Exception throw from native");
//                e.printStackTrace();
//            }
//        }
//    }
//
//    /**
//     * Render page fragment on {@link Bitmap}.<br>
//     * Page must be opened before rendering.
//     * <p>
//     * Supported bitmap configurations:
//     * <ul>
//     * <li>ARGB_8888 - best quality, high memory usage, higher possibility of OutOfMemoryError
//     * <li>RGB_565 - little worse quality, twice less memory usage
//     * </ul>
//     */
//    public void renderPageBitmap(PdfDocument doc, Bitmap bitmap, int pageIndex,
//                                 int startX, int startY, int drawSizeX, int drawSizeY) {
//        renderPageBitmap(doc, bitmap, pageIndex, startX, startY, drawSizeX, drawSizeY, false);
//    }
//
//    /**
//     * Render page fragment on {@link Bitmap}. This method allows to render annotations.<br>
//     * Page must be opened before rendering.
//     * <p>
//     * For more info see {@link PdfiumCore#renderPageBitmap(PdfDocument, Bitmap, int, int, int, int, int)}
//     */
//    public void renderPageBitmap(PdfDocument doc, Bitmap bitmap, int pageIndex,
//                                 int startX, int startY, int drawSizeX, int drawSizeY,
//                                 boolean renderAnnot) {
//        synchronized (lock) {
//            try {
//                nativeRenderPageBitmap(doc.mNativePagesPtr.get(pageIndex), bitmap, mCurrentDpi,
//                        startX, startY, drawSizeX, drawSizeY, renderAnnot);
//            } catch (NullPointerException e) {
//                Log.e(TAG, "mContext may be null");
//                e.printStackTrace();
//            } catch (Exception e) {
//                Log.e(TAG, "Exception throw from native");
//                e.printStackTrace();
//            }
//        }
//    }
//
//    /**
//     * Map page coordinates to device screen coordinates
//     *
//     * @param doc       pdf document
//     * @param pageIndex index of page
//     * @param startX    left pixel position of the display area in device coordinates
//     * @param startY    top pixel position of the display area in device coordinates
//     * @param sizeX     horizontal size (in pixels) for displaying the page
//     * @param sizeY     vertical size (in pixels) for displaying the page
//     * @param rotate    page orientation: 0 (normal), 1 (rotated 90 degrees clockwise),
//     *                  2 (rotated 180 degrees), 3 (rotated 90 degrees counter-clockwise)
//     * @param pageX     X value in page coordinates
//     * @param pageY     Y value in page coordinate
//     * @return mapped coordinates
//     */
//    public Point mapPageCoordsToDevice(PdfDocument doc, int pageIndex, int startX, int startY, int sizeX,
//                                       int sizeY, int rotate, double pageX, double pageY) {
//        long pagePtr = doc.mNativePagesPtr.get(pageIndex);
//        return nativePageCoordsToDevice(pagePtr, startX, startY, sizeX, sizeY, rotate, pageX, pageY);
//    }
//
//    /**
//     * @return mapped coordinates
//     * @see PdfiumCore#mapPageCoordsToDevice(PdfDocument, int, int, int, int, int, int, double, double)
//     */
//    public RectF mapRectToDevice(PdfDocument doc, int pageIndex, int startX, int startY, int sizeX,
//                                 int sizeY, int rotate, RectF coords) {
//
//        Point leftTop = mapPageCoordsToDevice(doc, pageIndex, startX, startY, sizeX, sizeY, rotate,
//                coords.left, coords.top);
//        Point rightBottom = mapPageCoordsToDevice(doc, pageIndex, startX, startY, sizeX, sizeY, rotate,
//                coords.right, coords.bottom);
//        return new RectF(leftTop.x, leftTop.y, rightBottom.x, rightBottom.y);
//    }
//}

File: DocumentsSDK/DocumentPdf/src/main/java/com/proxy/service/document/pdf/core/PdfiumCore.kt
package com.proxy.service.document.pdf.core

import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.RectF
import android.view.Surface
import com.proxy.service.document.base.pdf.info.CatalogueData
import com.proxy.service.document.base.pdf.info.LinkData
import com.proxy.service.document.base.pdf.info.MetaData
import com.proxy.service.document.base.pdf.info.PageSize
import com.proxy.service.document.pdf.constants.Constants

/**
 * @author: cangHX
 * @data: 2025/5/2 08:19
 * @desc:
 */
class PdfiumCore {

    companion object {
        init {
            System.loadLibrary("c++_shared");
            System.loadLibrary("modpng");
            System.loadLibrary("modft2");
            System.loadLibrary("modpdfium");
            System.loadLibrary("csPdf");
        }

        private val core by lazy {
            PdfiumCore()
        }

        fun getInstance(): PdfiumCore {
            return core
        }
    }


    /*** *** *** *** *** *** *** 文档加载 *** *** *** *** *** *** ***/

    /**
     * 加载文档
     *
     * @param filePath  文件路径
     * @param password  密码
     * */
    external fun nativeOpenDocumentByPath(filePath: String, password: String?): OpenResult

    /**
     * 加载文档
     *
     * @param fd        文件描述符
     * @param password  密码
     * */
    external fun nativeOpenDocumentByFd(fd: Int, password: String?): OpenResult

    /**
     * 加载文档
     *
     * @param data      文件内容
     * @param password  密码
     * */
    external fun nativeOpenDocumentByMem(data: ByteArray, password: String?): OpenResult

    /**
     * 释放对应文档
     *
     * @param doc_hand 文档指针
     */
    external fun nativeCloseDocument(doc_hand: Long)


    /*** *** *** *** *** *** *** 文档基础信息 *** *** *** *** *** *** ***/

    /**
     * 获取对应文档内的页面数量
     *
     * @param doc_hand 文档指针
     */
    external fun nativeGetPageCount(doc_hand: Long): Int

    /**
     * 获取对应文档说明信息
     *
     * @param doc_hand 文档指针
     * @param tag      标签
     */
    private external fun nativeGetDocumentMetaText(doc_hand: Long, tag: String): String


    /*** *** *** *** *** *** *** 文档目录 *** *** *** *** *** *** ***/

    /**
     * 获取对应文档目录的第一个子目录指针
     *
     * @param doc_hand       文档指针
     * @param catalogue_hand 文档目录指针
     */
    private external fun nativeGetFirstChildBookmark(doc_hand: Long, catalogue_hand: Long?): Long?

    /**
     * 获取对应文档目录的下一个目录指针
     *
     * @param doc_hand       文档指针
     * @param catalogue_hand 文档目录指针
     */
    private external fun nativeGetNextChildBookmark(doc_hand: Long, catalogue_hand: Long): Long?

    /**
     * 获取对应文档目录标题
     *
     * @param catalogue_hand 文档目录指针
     */
    private external fun nativeGetBookmarkTitle(catalogue_hand: Long): String?

    /**
     * 获取文档目录对应的页码
     *
     * @param doc_hand       文档指针
     * @param catalogue_hand 文档目录指针
     */
    private external fun nativeGetBookmarkDestIndex(doc_hand: Long, catalogue_hand: Long): Long


    /*** *** *** *** *** *** *** 文档页面加载 *** *** *** *** *** *** ***/

    /**
     * 加载对应文档页面, 并获取页面指针
     *
     * @param doc_hand  文档指针
     * @param pageIndex 页面位置
     */
    external fun nativeLoadPage(doc_hand: Long, pageIndex: Int): Long

    /**
     * 批量加载对应文档页面, 并获取页面指针
     *
     * @param doc_hand  文档指针
     * @param fromIndex 页面开始位置
     * @param toIndex   页面结束位置
     */
    external fun nativeLoadPages(doc_hand: Long, fromIndex: Int, toIndex: Int): LongArray

    /**
     * 关闭文档对应页面
     *
     * @param page_hand 页面指针
     */
    external fun nativeClosePage(page_hand: Long)

    /**
     * 批量关闭文档对应页面
     *
     * @param pages_hands 页面指针
     */
    external fun nativeClosePages(pages_hands: LongArray)


    /*** *** *** *** *** *** *** 页面尺寸信息 *** *** *** *** *** *** ***/

    /**
     * 获取文档对应页面宽度, 单位: 像素
     *
     * @param page_hand 页面指针
     * @param dpi       像素密度
     * */
    external fun nativeGetPageWidthPixel(page_hand: Long, dpi: Int): Int

    /**
     * 获取文档对应页面高度, 单位: 像素
     *
     * @param page_hand 页面指针
     * @param dpi       像素密度
     * */
    external fun nativeGetPageHeightPixel(page_hand: Long, dpi: Int): Int

    /**
     * 获取文档对应页面宽度, 单位: 文档点
     *
     * @param page_hand 页面指针
     * */
    external fun nativeGetPageWidthPoint(page_hand: Long): Int

    /**
     * 获取文档对应页面高度, 单位: 文档点
     *
     * @param page_hand 页面指针
     * */
    external fun nativeGetPageHeightPoint(page_hand: Long): Int

    /**
     * 通过页码获取对应文档页面尺寸, 单位: 像素
     *
     * @param doc_hand  文档指针
     * @param pageIndex 文档页码
     * @param dpi       像素密度
     * */
    private external fun nativeGetPageSizeByIndex(
        doc_hand: Long,
        pageIndex: Int,
        dpi: Int
    ): PageSize


    /*** *** *** *** *** *** *** 页面超链 *** *** *** *** *** *** ***/

    /**
     * 获取文档对应页面的全部超链接信息
     *
     * @param page_hand 页面指针
     * */
    private external fun nativeGetPageLinks(page_hand: Long): LongArray

    /**
     * 获取超链接对应的目标页码
     *
     * @param doc_hand  页面指针
     * @param link_hand 超链接指针
     * */
    private external fun nativeGetDestPageIndex(doc_hand: Long, link_hand: Long): Int?

    /**
     * 获取超链接对应的目标 uri
     *
     * @param doc_hand  页面指针
     * @param link_hand 超链接指针
     * */
    private external fun nativeGetLinkURI(doc_hand: Long, link_hand: Long): String?

    /**
     * 获取超链接对应的坐标区域
     *
     * @param link_hand 超链接指针
     * */
    private external fun nativeGetLinkRectF(link_hand: Long): RectF?


    /*** *** *** *** *** *** *** 页面渲染-整体渲染 *** *** *** *** *** *** ***/

    /**
     * 渲染页面内容到 surface
     *
     * @param page_hand     页面指针
     * @param surface       surface
     * @param dpi           像素密度
     * @param startX        渲染区域的开始坐标, 相对于 bitmap 左上角的 x 轴位置
     * @param startY        渲染区域的开始坐标, 相对于 bitmap 左上角的 y 轴位置
     * @param width         渲染区域的宽度
     * @param height        渲染内容的高度
     * @param renderAnnot   是否渲染批注（高亮、下划线、便签等）
     * @param viewBgColor   视图背景色, 格式为：0xAARRGGBB
     * @param pageBgColor   页面背景色, 格式为：0xAARRGGBB
     * */
    external fun nativeRenderPageToSurface(
        page_hand: Long,
        surface: Surface,
        dpi: Int,
        startX: Int,
        startY: Int,
        width: Int,
        height: Int,
        renderAnnot: Boolean,
        viewBgColor: Long,
        pageBgColor: Long
    )

    /**
     * 渲染页面内容到 bitmap
     *
     * @param page_hand     页面指针
     * @param bitmap        bitmap
     * @param dpi           像素密度
     * @param startX        渲染区域的开始坐标, 相对于 bitmap 左上角的 x 轴位置
     * @param startY        渲染区域的开始坐标, 相对于 bitmap 左上角的 y 轴位置
     * @param width         渲染区域的宽度
     * @param height        渲染内容的高度
     * @param renderAnnot   是否渲染批注（高亮、下划线、便签等）
     * @param viewBgColor   视图背景色, 格式为：0xAARRGGBB
     * @param pageBgColor   页面背景色, 格式为：0xAARRGGBB
     * */
    external fun nativeRenderPageToBitmap(
        page_hand: Long,
        bitmap: Bitmap,
        dpi: Int,
        startX: Int,
        startY: Int,
        width: Int,
        height: Int,
        renderAnnot: Boolean,
        viewBgColor: Long,
        pageBgColor: Long
    )


    /*** *** *** *** *** *** *** 页面渲染-渐进式渲染 *** *** *** *** *** *** ***/


    /*** *** *** *** *** *** *** 通用能力 *** *** *** *** *** *** ***/

    /**
     * 页面坐标映射为屏幕坐标
     *
     * @param pagePtr 页面指针
     * @param startX  显示区域在设备坐标中的左侧像素位置
     * @param startY  显示区域在设备坐标中的顶部像素位置
     * @param sizeX   显示区域的宽度(单位：像素)
     * @param sizeY   显示区域的高度(单位：像素)
     * @param rotate  页面方向：0（正常）, 1（顺时针旋转90度）, 2（顺时针旋转180度）, 3（顺时针旋转270度）
     * @param pageX   页面坐标中的X值
     * @param pageY   页面坐标中的Y值
     * @return 返回映射坐标
     */
    private external fun nativePageCoordsToDevice(
        pagePtr: Long,
        startX: Int,
        startY: Int,
        sizeX: Int,
        sizeY: Int,
        rotate: Int,
        pageX: Double,
        pageY: Double
    ): Point?


    /**************************************************/


    /**
     * 获取对应文档说明信息
     *
     * @param doc_hand 文档指针
     */
    fun getDocumentMeta(doc_hand: Long): MetaData {
        val meta = MetaData()
        meta.title = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_TITLE)
        meta.author = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_AUTHOR)
        meta.subject = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_SUBJECT)
        meta.keywords = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_KEYWORDS)
        meta.creator = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_CREATOR)
        meta.producer = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_PRODUCER)
        meta.creationDate =
            core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_CREATION_DATE)
        meta.modDate = core.nativeGetDocumentMetaText(doc_hand, Constants.DOC_META_MOD_DATE)
        return meta
    }

    /**
     * 获取对应文档目录信息
     *
     * @param doc_hand 文档指针
     */
    fun getDocumentCatalogue(doc_hand: Long): ArrayList<CatalogueData> {
        val list = ArrayList<CatalogueData>()
        val hand = core.nativeGetFirstChildBookmark(doc_hand, null)
        if (hand != null) {
            queryDocumentCatalogue(list, core, doc_hand, hand)
        }
        return list
    }

    private fun queryDocumentCatalogue(
        list: ArrayList<CatalogueData>,
        core: PdfiumCore,
        doc_hand: Long,
        catalogue_hand: Long
    ) {
        val catalogueData = CatalogueData()
        catalogueData.hand = catalogue_hand
        catalogueData.title = core.nativeGetBookmarkTitle(catalogue_hand)
        catalogueData.pageIndex = core.nativeGetBookmarkDestIndex(doc_hand, catalogue_hand)
        list.add(catalogueData)

        val child_catalogue_hand = core.nativeGetFirstChildBookmark(doc_hand, catalogue_hand)
        if (child_catalogue_hand != null) {
            queryDocumentCatalogue(
                catalogueData.children,
                core,
                doc_hand,
                child_catalogue_hand
            )
        }

        val next_catalogue_hand = core.nativeGetNextChildBookmark(doc_hand, catalogue_hand)
        if (next_catalogue_hand != null) {
            queryDocumentCatalogue(list, core, doc_hand, next_catalogue_hand)
        }
    }

    /**
     * 获取对应页面的超链接信息
     *
     * @param doc_hand  文档指针
     * @param page_hand 页面指针
     */
    fun getPageLinks(doc_hand: Long, page_hand: Long): ArrayList<LinkData> {
        val list = ArrayList<LinkData>()
        val link_hands = nativeGetPageLinks(page_hand)

        link_hands.forEach {
            val rectf = nativeGetLinkRectF(it) ?: return@forEach

            val index = nativeGetDestPageIndex(doc_hand, it)
            val uri = nativeGetLinkURI(doc_hand, it)

            if (index != null || uri != null) {
                return@forEach
            }

            list.add(LinkData(rectf, index, uri))
        }
        return list
    }
}